#  Drawing polygon to create training data <img align="right" src="image/gistda.png" width="200" height="200">

 - [Display interactive map](#Display-interactive-map)
 - [Display output from drawing](#Display-output-from-drawing)
 - [Add new column](#Add-new-column)
 - [Merge data](#Merge-data)
 - [Convert geographic coordinate system](#convert-geographic-coordinate-system) 
 
<table class="ee-notebook-buttons" align="left">
    <td><a target="_parent"  href="https://nbviewer.org/github/Pathakorn40/rice-detection/blob/main/Tutorial/Drawing_polygon.ipynb"><img width=26px src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Jupyter_logo.svg/883px-Jupyter_logo.svg.png" />Notebook Viewer</a></td>
</table>

### Display interactive map


This case study draw the polygon to create training data for ML classification.
- Draw polygons from the tools in left side and export file by `Export` buttom on the right side .
- This case study, we catagorize training data into three classes that compose of urban area, water area and green area.
- The output will save in GeoJson file format.

We draw the polygon in 3 files name: green_area_sawan.geojson, water_sawan.geojson,and urban_sawan.geojson.

In [1]:
from dream_river.plotimg import show_map

Map = show_map()
Map

## Display output from drawing

In [2]:
from dream_river.plotimg import show_vector

# get path of files
path = ['training/green_area_sawan.geojson','training/water_sawan.geojson','training/urban_sawan.geojson']
show_vector(path)

### Data preparation

In [3]:
import geopandas as gpd

# open file in geopandas dataframe
gdf = gpd.read_file('training/green_area_sawan.geojson')

# to see 5 first rows
gdf.head()

class                                           geometry
0    111  POLYGON ((100.12068 15.71600, 100.12068 15.716...
1    111  POLYGON ((100.13291 15.71772, 100.13291 15.718...
2    111  POLYGON ((100.13212 15.71351, 100.13212 15.713...
3    111  POLYGON ((100.13496 15.71275, 100.13504 15.712...
4    111  POLYGON ((100.13395 15.71272, 100.13444 15.712...

As you can see the table above, It do not have any unique value for identify the class type. 

### Add new column 
- to categorize the type of polygon files for creating training data
- first, input the name of new column. Then assign the code value to identify the type of that file.

In this case study, we create "class" column and add value into that column follow file name to categorize the type of data like: 

green_area = 111

water = 112

urban = 113

In [4]:
from dream_river.plotimg import geojson_add_Newcol

# func to add new colume and assign value
geojson_add_Newcol('training/green_area_sawan.geojson')

num of rows:  20


Enter the name of the new column:  class
Enter the value for the new column:  111


20

In [5]:
# func to add new colume and assign value
geojson_add_Newcol('training/water_sawan.geojson')

num of rows:  14


Enter the name of the new column:  class
Enter the value for the new column:  112


14

In [6]:
# func to add new colume and assign value
geojson_add_Newcol('training/urban_sawan.geojson')

num of rows:  18


Enter the name of the new column:  class
Enter the value for the new column:  113


18

### Explore output

In [7]:
import geopandas as gpd

gdf = gpd.read_file('training/urban_sawan.geojson')

# gdf.head(50)
gdf.head()

class                                           geometry
0    113  POLYGON ((100.13354 15.70495, 100.13372 15.705...
1    113  POLYGON ((100.13938 15.70688, 100.13843 15.704...
2    113  POLYGON ((100.14057 15.70553, 100.13964 15.704...
3    113  POLYGON ((100.14808 15.71282, 100.14808 15.714...
4    113  POLYGON ((100.15108 15.71813, 100.15100 15.717...

In [8]:
# check coordinate system of data
gdf.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

### Merge data
Note : It will take a few secound second to create file.

In [9]:
from dream_river.geobox import merge

# assige input path
file =['training/green_area_sawan.geojson','training/urban_sawan.geojson','training/water_sawan.geojson']

# assign output file path
output = 'training/training_data.geojson'

# merge function
merge(file,output)

### Explore output

In [10]:
gdf = gpd.read_file('training/training_data.geojson')

gdf.head(25)
# gdf.head()

class                                           geometry
0     111  POLYGON ((100.12068 15.71600, 100.12068 15.716...
1     111  POLYGON ((100.13291 15.71772, 100.13291 15.718...
2     111  POLYGON ((100.13212 15.71351, 100.13212 15.713...
3     111  POLYGON ((100.13496 15.71275, 100.13504 15.712...
4     111  POLYGON ((100.13395 15.71272, 100.13444 15.712...
5     111  POLYGON ((100.12376 15.70586, 100.12376 15.706...
6     111  POLYGON ((100.12419 15.70799, 100.12419 15.708...
7     111  POLYGON ((100.12467 15.70932, 100.12467 15.709...
8     111  POLYGON ((100.12602 15.70986, 100.12602 15.710...
9     111  POLYGON ((100.13291 15.70694, 100.13291 15.707...
10    111  POLYGON ((100.13556 15.70449, 100.13529 15.703...
11    111  POLYGON ((100.11181 15.70125, 100.11181 15.702...
12    111  POLYGON ((100.10953 15.70873, 100.10953 15.709...
13    111  POLYGON ((100.10743 15.71212, 100.10743 15.712...
14    111  POLYGON ((100.11135 15.71299, 100.11135 15.713...
15    111  POLYGON ((100.10132 15.71485, 100.10132 15.715...
16    111  POLYGON ((100.12456 15.71932, 100.12456 15.720...
17    111  POLYGON ((100.12250 15.72322, 100.12250 15.723...
18    111  POLYGON ((100.12178 15.72668, 100.12178 15.726...
19    111  POLYGON ((100.09770 15.69744, 100.09770 15.697...
20    113  POLYGON ((100.13354 15.70495, 100.13372 15.705...
21    113  POLYGON ((100.13938 15.70688, 100.13843 15.704...
22    113  POLYGON ((100.14057 15.70553, 100.13964 15.704...
23    113  POLYGON ((100.14808 15.71282, 100.14808 15.714...
24    113  POLYGON ((100.15108 15.71813, 100.15100 15.717...

In [11]:
from dream_river.plotimg import show_vector

path = ['training/training_data.geojson']
show_vector(path)

### Convert geographic coordinate system
- This case study convert the CRS from EPSG:4326(degree) to EPSG:32647(meter)

In [12]:
# check coordinate system of data
gdf.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

In [13]:
from dream_river.geobox import convert_geojson_CRS

# set input output path
input_path = 'training/training_data.geojson'
output_path = 'training/training_32647.geojson'

# function to convert CRS
convert_geojson_CRS(input_path, output_path, crs=32647)

In [14]:
# Load input data shapefile
input1 = gpd.read_file(output_path)

# checl crs
input1.crs

<Derived Projected CRS: EPSG:32647>
Name: WGS 84 / UTM zone 47N
Axis Info [cartesian]:
- E[east]: Easting (metre)
- N[north]: Northing (metre)
Area of Use:
- name: Between 96°E and 102°E, northern hemisphere between equator and 84°N, onshore and offshore. China. Indonesia. Laos. Malaysia - West Malaysia. Mongolia. Myanmar (Burma). Russian Federation. Thailand.
- bounds: (96.0, 0.0, 102.0, 84.0)
Coordinate Operation:
- name: UTM zone 47N
- method: Transverse Mercator
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich